File: modules/photo-vision/android/src/main/java/expo/modules/photovision/PhotoVisionModule.kt
package expo.modules.photovision

//
//  Step 3 — Expo Modules API 브릿지 (Android)
//
//  iOS(PhotoVisionModule.swift)와 동일한 모듈명/함수 시그니처를 제공.
//  단일 스레드 Executor에서 1장씩 순차 처리하여 메모리 피크를 억제한다.
//

import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.util.concurrent.Executors

class PhotoVisionModule : Module() {

    // 사진 분석 전용 백그라운드 스레드 (발열/메모리 제어)
    private val executor = Executors.newSingleThreadExecutor()

    override fun definition() = ModuleDefinition {
        Name("PhotoVision")

        // Android에는 OS 미학 점수 API가 없음
        Constants("aestheticsAvailable" to false)

        AsyncFunction("analyzePhotos") { uris: List<String>, promise: Promise ->
            executor.execute {
                try {
                    val results = uris.map { PhotoVisionAnalyzer.analyze(it) }
                    promise.resolve(results)
                } catch (e: Exception) {
                    promise.reject("ANALYZE_FAILED", e.message ?: "사진 분석 실패", e)
                }
            }
        }
    }
}

File: modules/photo-vision/ios/PhotoVisionModule.swift
//
//  PhotoVisionModule.swift
//  Step 2 — Expo Modules API 브릿지
//
//  JS의 analyzePhotos(uris) 호출을 백그라운드 큐에서 1장씩 처리하고,
//  각 장을 autoreleasepool로 감싸 메모리를 즉시 회수한다(OOM 방지).
//

import ExpoModulesCore

public class PhotoVisionModule: Module {
    public func definition() -> ModuleDefinition {
        Name("PhotoVision")

        // 이 모듈이 동작 가능한 환경인지 (iOS 18+ 미학 점수 가용 여부 포함)
        Constants([
            "aestheticsAvailable": {
                if #available(iOS 18.0, *) { return true } else { return false }
            }()
        ])

        // file:// 썸네일 경로 배열을 받아 품질 지표 배열을 돌려준다.
        AsyncFunction("analyzePhotos") { (uris: [String], promise: Promise) in
            DispatchQueue.global(qos: .utility).async {
                var results: [[String: Any]] = []
                results.reserveCapacity(uris.count)
                for uri in uris {
                    autoreleasepool {
                        results.append(PhotoVisionAnalyzer.analyze(uri: uri))
                    }
                }
                promise.resolve(results)
            }
        }
    }
}
